src/abe/dippe.c

```diff
@@ -0,0 +1,613 @@
+    return i ? (((int)log10((double)i))+1) : 1;
+}
+
+/**
+ * Hashes the given buffer to an element of G2
+ *
+ * @param out A pointer to a ECP2_BN254 struct; The resulting group element
+ * @param in_buf Buffer that is about to be hashed
+ * @param in_buf_size Size of the buffer
+ */
+void cfe_dippe_hash_G2(ECP2_BN254 *out, char* in_buf, int in_buf_size) {
+    // rounds
+    int r = (in_buf_size / MODBYTES_256_56) + ((in_buf_size % MODBYTES_256_56) != 0);
+    // hash buffer
+    char buf[MODBYTES_256_56];
+
+    octet tmp;
+    tmp.val = (char*)&buf;
+    tmp.len = MODBYTES_256_56;
+    tmp.max = MODBYTES_256_56;
+
+    ECP2_BN254_inf(out);
+    ECP2_BN254 tmp_g2;
+
+    for (int i=0; i<r; i++) {
+        for (int j=0; j<MODBYTES_256_56; j++) {
+            if (((i*MODBYTES_256_56)+j) < in_buf_size) buf[j] = in_buf[(i*MODBYTES_256_56)+j]; else buf[j] = '\0';
+            if ((((i*MODBYTES_256_56)+j+1) % MODBYTES_256_56) == 0) {
+                ECP2_BN254_mapit(&tmp_g2, &tmp);
+                ECP2_BN254_add(out, &tmp_g2);
+            }
+        }
+    }
+}
+
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+/**
+ * Clears the scheme and frees allocated memory
+ *
+ * @param dippe A pointer to a cfe_dippe struct
+ */
+void cfe_dippe_free(cfe_dippe *dippe) {
+    mpz_clear(dippe->p);
+    free(dippe->g1_A);
+    free(dippe->g1_UA);
+}
+
+/**
+ * Clears the public key and frees allocated memory
+ *
+ * @param pk A pointer to a cfe_dippe_pub_key struct
+ */
+void cfe_dippe_pub_key_free(cfe_dippe_pub_key *pk) {
+	free(pk->gt_alpha_A);
+	free(pk->g1_W_A);
+}
+
+/**
+ * Clears the secret key and frees allocated memory
+ *
+ * @param sk A pointer to a cfe_dippe_sec_key struct
+ */
+void cfe_dippe_sec_key_free(cfe_dippe_sec_key *sk) {
+	mpz_clear(sk->sigma);
+	cfe_vec_free(&(sk->alpha));
+	cfe_vec_free(&(sk->W));
+}
+
+/**
+ * Clears the ciphertext and frees allocated memory
+ *
+ * @param cipher A pointer to a cfe_dippe_cipher struct
+ */
+void cfe_dippe_cipher_free(cfe_dippe_cipher *cipher) {
+    free(cipher->C0);
+    free(cipher->Ci);
+}
+
+/**
+ * Clears the user secret key and frees allocated memory
+ *
+ * @param usk A pointer to a cfe_dippe_user_secret_key struct
+ */
+void cfe_dippe_user_secret_key_free(cfe_dippe_user_secret_key *usk) {
+    free(usk->Ki);
+}
+
+/**
+ * Initializes the DIPPE scheme and generates the global parameters
+ *
+ * @param dippe A pointer to a cfe_dippe struct
+ * @param assump_size The size of the underlying assumption
+ */
+void cfe_dippe_setup(cfe_dippe *dippe, unsigned int assump_size) {
```

Comment:
There is no need to copy these comments from the header files. Only the header files are used for creating the documentation and if you're using an IDE, you can quickly jump to a function's definition from its implementation. I don't necessarily recommend deleting all of these comments in the `.c` file, maybe you can keep just the descriptions (up to you).

Reply:
You're right, I removed the comment blocks from the c-files.